server/src/main/java/org/elasticsearch/repositories/FinalizeSnapshotContext.java

```diff
@@ -106,8 +108,21 @@ public ClusterState updatedClusterState(ClusterState state) {
         final ClusterState updatedState = SnapshotsService.stateWithoutSnapshot(state, snapshotInfo.snapshot(), updatedShardGenerations);
         // Now that the updated cluster state may have changed in-progress shard snapshots' shard generations to the latest shard
         // generation, let's mark any now unreferenced shard generations as obsolete and ready to be deleted.
+
+        final Collection<IndexId> deletedIndices = updatedShardGenerations.deletedIndices.indices();
         obsoleteGenerations.set(
-            SnapshotsInProgress.get(updatedState).obsoleteGenerations(snapshotInfo.repository(), SnapshotsInProgress.get(state))
+            SnapshotsInProgress.get(updatedState)
+                .obsoleteGenerations(snapshotInfo.repository(), SnapshotsInProgress.get(state))
+                .entrySet()
+                .stream()
+                // We want to keep both old and new generations for deleted indices, so we filter them out here to avoid deletion.
+                // We need the old generations because they are what get recorded in the RepositoryData.
+                // We also need the new generations a future finalization may build upon them. It may ends up not being used at all
+                // when current batch of in-progress snapshots are completed and no new index of the same name is created.
+                // That is also OK. It means we have some redundant shard generations in the repository and they will be deleted
+                // when a snapshot deletion runs.
+                .filter(e -> deletedIndices.contains(e.getKey().index()) == false)
+                .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue))
```

Comment:
If we do decide to keep both old and new gens, I'd rather we pushed this down into the implementation of `obsoleteGenerations()` rather than building a whole new map afresh via a stream as proposed here. This stuff is somewhat performance-critical, and we can filter out the deleted indices way more cheaply if we do it in `obsoleteGenerations()`.